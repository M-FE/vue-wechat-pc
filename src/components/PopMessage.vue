<script>
export default {
    props: {
        placement: {
            type: String,
            default: 'left' /* left/right */
        },
        maxWidth: {
            type: [String, Number],
            default: null
        },
        backgroundColor: {
            type: String,
            default: '#fff'
        }
    },

    computed: {
        className() {
            if (this.placement === 'right') {
                return 'pop-right';
            }

            return 'pop-left';
        },
        style() {
            if (!this.maxWidth) {
                return {};
            }

            if (typeof this.maxWidth === 'number') {
                return { 'max-width': this.maxWidth + 'px' };
            }

            return { 
                'max-width': this.maxWidth,
                backgroundColor: this.backgroundColor
            };
        }
    },

    render() {
        return (
            <div data-test="pop-message" style={this.style} class={[
                this.className,
                'pop-message'
            ]}>{ this.$slots.default }</div>
        );
    }
};
</script>

<style lang="scss">
@import '../assets/scss/var.scss';

.pop-message {
    position: relative;
    display: inline-block;
    border: 1px solid $color-6;
    border-radius: $radius;
    padding: 6px 12px;
    font-size: $text-base;
    line-height: 1.5;
    min-width: 30px;
    background-color: #fff;

    &:after {
        position: absolute;
        top: 10px;
        width: 6px;
        height: 6px;
        border: 1px solid $color-6;
        transform: rotate(45deg);
        background-color: #fff;
        content: '';
    }

    &.pop-left {
        &:after {
            left: -4px;
            border-width: 0 0 1px 1px;
        }
    }

    &.pop-right {
        &:after {
            right: -4px;
            border-width: 1px 1px 0 0;
        }
    }
}
</style>
