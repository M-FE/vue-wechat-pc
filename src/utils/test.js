/**
 * 在wrapper中查找data-test为name的元素
 * @param {*} wrapper vue-test-utils的wrapper
 * @param {string} name data-test的属性值
 * @return {*}
 */
export const findFromWrapper = (wrapper, name) => {
    return wrapper.findAll(`[data-test=${name}]`);
}
