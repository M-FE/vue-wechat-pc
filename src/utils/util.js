export const DEFAULT_USER_PROPS = {
    id: 'id',
    avatar: 'avatar',
    name: 'name',
    time: 'time',
    summary: 'summary'
};

export const DEFAULT_MESSAGE_PROPS = {
    id: 'id',
    content: 'content',
    time: 'time',
    other: 'other'
};

/**
 * 获取用于User的PropNames
 * @param {*} props props
 * @return {object}
 */
export const userPropNames = (props) => {
    return {
        ...DEFAULT_USER_PROPS,
        ...props
    };
}

export const messagePropNames = (props) => {
    return {
        ...DEFAULT_MESSAGE_PROPS,
        ...props
    };
}
