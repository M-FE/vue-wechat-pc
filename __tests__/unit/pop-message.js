import { shallowMount } from '@vue/test-utils';
import PopMessage from '@/components/PopMessage';
import { findFromWrapper } from '@/utils/test';

describe('Component - PopMessage', () => {
    it('应该存在一个默认插槽', () => {
        const wrapper = shallowMount(PopMessage, {
            slots: {
                default: '<div data-test="pop-desc"></div>'
            }
        });

        const popDefault = findFromWrapper(wrapper, 'pop-desc');

        expect(popDefault.length).toBe(1);
    });

    it('不传入placement时，默认有一个classname是pop-left', () => {
        const wrapper = shallowMount(PopMessage);
        const popMessage = findFromWrapper(wrapper, 'pop-message').at(0);
        expect(popMessage.classes('pop-left')).toBe(true);
    });

    it('placement为right时，存在一个classname为pop-right', () => {
        const wrapper = shallowMount(PopMessage, {
            propsData: { placement: 'right' }
        });
        const popMessage = findFromWrapper(wrapper, 'pop-message').at(0);
        expect(popMessage.classes('pop-right')).toBe(true);
    });

    it('使用maxWidth控制最长宽度', () => {
        const wrapper = shallowMount(PopMessage, {
            propsData: {
                maxWidth: 200
            }
        });
        const popMessage = findFromWrapper(wrapper, 'pop-message').at(0);
        const styles = popMessage.attributes('style') || '';

        expect(/max-width: ?200px;?/.test(styles || '')).toBeTruthy();
    });

    it('生成快照，确定结构', () => {
        const wrapper = shallowMount(PopMessage, {
            propsData: {
                placement: 'right',
                maxWidth: 200
            },
            slots: {
                default: '<div>Hello</div>'
            }
        });

        expect(wrapper).toMatchSnapshot();
    });
});
