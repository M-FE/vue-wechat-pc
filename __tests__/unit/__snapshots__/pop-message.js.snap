// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - PopMessage 生成快照，确定结构 1`] = `
<div data-test="pop-message" class="pop-right pop-message" style="max-width: 200px;">
  <div>Hello</div>
</div>
`;
